src/EventStore.Core/ProcessTracker.cs

```diff
@@ -0,0 +1,226 @@
+	private readonly DateTime _startTime;
+	private readonly TimeSpan _startCpuUsage;
+	private readonly string _dbPath;
+	private readonly PerfCounterHelper _perfCounter;
+
+	public Histogram<float> SysCPU { get; set; }
+	public Histogram<long> SysFreeMem { get; set; }
+	public Histogram<long> SysTotalMem { get; set; }
+
+	public Histogram<float> ProcessCPU { get; set; }
+
+	public Counter<long> DiskIOReadBytes { get; set; }
+	public Counter<long> DiskIOReadOps { get; set; }
+	public Counter<long> DiskIOWrittenBytes { get; set; }
+	public Counter<long> DiskIOWriteOps { get; set; }
+
+	public Histogram<long> DriveTotalBytes { get; set; }
+	public Histogram<long> DriveUsedBytes { get; set; }
+
+	public Histogram<float> GcAllocationSpeed { get; set; }
+	public Histogram<float> GcFragmentation { get; set; }
+	public Histogram<long> GcGen0Size { get; set; }
+	public Histogram<long> GcGen0ItemsCount { get; set; }
+	public Histogram<long> GcGen1Size { get; set; }
+	public Histogram<long> GcGen1ItemsCount { get; set; }
+	public Histogram<long> GcGen2Size { get; set; }
+	public Histogram<long> GcGen2ItemsCount { get; set; }
+	public Histogram<long> GcLohSize { get; set; }
+	public Histogram<float> GcTimeInGc { get; set; }
+	public Histogram<long> GcTotalBytesInHeaps { get; set; }
+
+	public ProcessTracker(IClock clock, HostStat.HostStat stats, string dbPath) {
+		_clock = clock;
+		_stats = stats;
+		_dbPath = dbPath;
+		_perfCounter = new PerfCounterHelper(_log);
+		_startTime = DateTime.Now;
+		_startCpuUsage = Process.GetCurrentProcess().TotalProcessorTime;
+	}
+
+	private T Bookkeepping<T>(Func<T> func) {
+		var now = _clock.Now;
+		if (_lastTime != Instant.FromSeconds(0) && now.ElapsedSecondsSince(_lastTime) < 5)
+			return func();
+
+		_lastTime = now;
+
+		if (OS.OsFlavor == OsFlavor.Linux || OS.OsFlavor == OsFlavor.MacOS) {
+			_loadAverages = _stats.GetLoadAverages();
+		}
+
+		_esDriveInfo = EsDriveInfo.FromDirectory(_dbPath, _log);
+		_diskIo = DiskIo.GetDiskIo(Environment.ProcessId, _log);
+		_process = Process.GetCurrentProcess();
+
+		return func();
+	}
+
+	public IEnumerable<Measurement<double>> CollectCpuAvg1M() {
+		return Bookkeepping<IEnumerable<Measurement<double>>>(() => {
+			return OS.OsFlavor == OsFlavor.Windows
+				? Array.Empty<Measurement<double>>()
+				: new[] { new Measurement<double>(_loadAverages.Average1m) };
+		});
+	}
+
+	public IEnumerable<Measurement<double>> CollectCpuAvg5M() {
+		return Bookkeepping<IEnumerable<Measurement<double>>>(() => {
+			return OS.OsFlavor == OsFlavor.Windows
+				? Array.Empty<Measurement<double>>()
+				: new[] { new Measurement<double>(_loadAverages.Average5m) };
+		});
+	}
+
+	public IEnumerable<Measurement<double>> CollectCpuAvg15M() {
+		return Bookkeepping<IEnumerable<Measurement<double>>>(() => {
+			return OS.OsFlavor == OsFlavor.Windows
+				? Array.Empty<Measurement<double>>()
+				: new[] { new Measurement<double>(_loadAverages.Average15m) };
+		});
+	}
```

Comment:
We can be a bit more careful with the allocations here. We're allocating (a) a Func and (b) an array each time this is called (although granted they're probably only gen0)

`CreateObservableUpDownCounter` has an overload where the func returns a `Measurement<>` rather than an `IEnumerable<Measurement<>>`, and perhaps we can do the OSFlavor check when creating the observableupdowncounter

which would leave us with something like this, which is a bit neater and avoids the array allocation

```
	public Measurement<double> CollectCpuAvg15M() {
		return Bookkeepping(() => new Measurement<double>(_loadAverages.Average15m));
	}
```

Reply:
probably to have a separate class to encapsulate the LoadAverages and decide whether to fetch new ones or return the current ones. and then the method can be something like

```
public Measurement<double> CollectCpuAvg15M() => new(_loadAverages.GetCurrent().Average15m);
```

Reply:
or if you like, the functionalish style works for me, something like:

```
public static class FuncDebouncer {
	public static Func<T> Debounce<T>(Func<T> func, TimeSpan timeout, IClock? clock = null) {
		clock ??= Clock.Instance;
		var seconds = timeout.TotalSeconds;
		Instant lastCalled = default;
		T? current = default;

		return () => {
			var now = clock.Now;
			if (lastCalled == default || now.ElapsedSecondsSince(lastCalled) > seconds) {
				current = func();
				lastCalled = now;
			}

			return current!;
		};
	}
}
```

```
		_getLoadAverages = FuncDebouncer.Debounce(_stats.GetLoadAverages, TimeSpan.FromSeconds(1.0));
```
```
	public Measurement<double> CollectCpuAvg1M() => new(_getLoadAverages().Average1m);
```

Reply:
this is implemented, fab :+1: